server/fishtest/rundb.py

```diff
@@ -288,45 +288,45 @@ def flush_all(self):
-                ):
-                    self.run_cache[r_id]["scavenge"] = now
-                    if self.scavenge(self.run_cache[r_id]["run"]):
-                        with self.run_cache_write_lock:
-                            self.runs.replace_one(
-                                {"_id": ObjectId(r_id)}, self.run_cache[r_id]["run"]
-                            )
-                if self.run_cache[r_id]["rtime"] < now - 300:
-                    del self.run_cache[r_id]
-            elif self.run_cache[r_id]["ftime"] < old:
-                old = self.run_cache[r_id]["ftime"]
-                oldest = r_id
-        # print(oldest)
-        if oldest is not None:
-            self.scavenge(self.run_cache[oldest]["run"])
-            self.run_cache[oldest]["scavenge"] = now
-            self.run_cache[oldest]["dirty"] = False
-            self.run_cache[oldest]["ftime"] = time.time()
-            self.run_cache_lock.release()  # Release the lock while writing
-            # print("SYNC")
-            with self.run_cache_write_lock:
-                self.runs.replace_one(
-                    {"_id": ObjectId(oldest)}, self.run_cache[oldest]["run"]
-                )
-            # start the timer when writing is done
+        try:
+            self.run_cache_lock.acquire()
+            now = time.time()
+            old = now + 1
+            oldest = None
+            for r_id in list(self.run_cache):
+                if not self.run_cache[r_id]["dirty"]:
+                    if not self.run_cache[r_id]["run"].get("finished", False) and (
+                        "scavenge" not in self.run_cache[r_id]
+                        or self.run_cache[r_id]["scavenge"] < now - 60
+                    ):
+                        self.run_cache[r_id]["scavenge"] = now
+                        if self.scavenge(self.run_cache[r_id]["run"]):
+                            with self.run_cache_write_lock:
+                                self.runs.replace_one(
+                                    {"_id": ObjectId(r_id)}, self.run_cache[r_id]["run"]
+                                )
+                    if self.run_cache[r_id]["rtime"] < now - 300:
+                        del self.run_cache[r_id]
+                elif self.run_cache[r_id]["ftime"] < old:
+                    old = self.run_cache[r_id]["ftime"]
+                    oldest = r_id
+            # print(oldest)
+            if oldest is not None:
+                self.scavenge(self.run_cache[oldest]["run"])
+                self.run_cache[oldest]["scavenge"] = now
+                self.run_cache[oldest]["dirty"] = False
+                self.run_cache[oldest]["ftime"] = time.time()
+                # print("SYNC")
+                with self.run_cache_write_lock:
+                    self.runs.replace_one(
+                        {"_id": ObjectId(oldest)}, self.run_cache[oldest]["run"]
+                    )
+        except:
+            print("Flush exception")
+        finally:
```

Comment:
DEV updated.
If I'm not wrong the PR, with this `finally:` section, simply releases the lock and starts the timer also after an exception. This looks good to me.

Reply:
Yes, that's one part of the PR. In addition I removed the writing without lock which I added some months ago to gain a potential speedup under heavy load. That is the root cause for the rare exception which I have seen twice (yesterday and a few weeks ago).

Reply:
I missed the `self.run_cache_lock.release()  # Release the lock while writing`.

Reply:
It might be that we need to grab the lock also on an heartbeat update, but I do not think that is needed. Let's wait to see if the log gets any "flush exception" messages in the future.